iroh-relay/src/server/clients.rs

```diff
@@ -85,8 +85,13 @@ impl Clients {
 
     /// Attempt to send a packet to client with [`NodeId`] `dst`
     pub(super) async fn send_packet(&self, dst: NodeId, data: Bytes, src: NodeId) -> Result<()> {
+        let mut res = None;
         if let Some(client) = self.0.clients.get(&dst) {
-            let res = client.try_send_packet(src, data);
+            res = Some(client.try_send_packet(src, data));
+        }
+        if let Some(res) = res {
+            // `process_result` may call `unregister`, which itself calls `self.0.clients`
+            // so we can't nest this under the above call to `self.0.clients`
```

Comment:
Oof, the size of this footgun!

I guess `&mut self` is a tool to fix that, but I assume that's difficult to do here because `Clients` needs to be `Clone` I assume, I didn't check how dearly that is needed but presumably that is what makes the routing scale across cores.  I don't have a useful suggestion without studying this deeper.